### ISSUE TITLE
Rancher Server is missing Thawte Premium Server CA

### ISSUE THREAD

**Issue by Seth-Karlo** [rancher/rancher#4293]
Rancher Version:
1.0.0

Docker Version:
Docker version 1.9.1, build 9894698

OS: 
CoreOS 991.2.0

Steps to Reproduce: 
Deploy rancher/server container, try to add a catalog over https with a certificate using SSLv3 or SSLv2

Results: 
Rancher server reports "There are no compatible templates" in the UI. 
Using docker log to see the logs will report similar to:
```
Apr 05 13:19:07 rancher-master-1 env[3601]: Cloning into './DATA/test'...
Apr 05 13:19:07 rancher-master-1 env[3601]: fatal: unable to access 'https://github.domain.com/arepton/test.git/': server certificate verification failed. CAfile: /etc/ssl/certs/ca-certificates.crt CRLfile: none
Apr 05 13:19:07 rancher-master-1 env[3601]: time="2016-04-05T13:19:07Z" level=error msg="Failed to clone the catalog from git err: exit status 128"
```
Logging onto the shell of the container using docker exec $id -it bash and running curl directly gives the error:
```
root@a172af1348d1:/# curl -Iv --cacert ca.crt https://github.domain.com/arepton/test.git
* Hostname was NOT found in DNS cache
*   Trying 195.66.90.140...
* Connected to github.schubergphilis.com (195.66.90.140) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: ca.crt
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* SSLv3, TLS handshake, Server hello (2):
* SSLv3, TLS handshake, CERT (11):
* SSLv3, TLS alert, Server hello (2):
* SSL certificate problem: unable to get local issuer certificate
* Closing connection 0
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: http://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
```
Passing through the ca.crt also did not work. However, the same commands on the hypervisor connected correctly.

Expected:
For rancher to successfully clone my repository.

Fix:

**tl;dr**: install  libgnutls-openssl27 and libssl1.0.0 inside the container

I originally followed the conversation at #3982 and noticed that using curl was causing issues. I was curious that it was working outside the container but not inside. On SSL certs which supported TLSv1 this would work fine, and I noticed that the container being used by rancher is using a very old version of openssl:
root@a172af1348d1:/# openssl version
OpenSSL 1.0.1f 6 Jan 2014
Vs my server:
core@rancher-master-1 ~ $ openssl version
OpenSSL 1.0.2g  1 Mar 2016

I connected to the container, ran apt-get upgrade openssl and then attempted to connect again. Now it worked.

One of these packages was upgraded which fixed it:
```root@a172af1348d1:/# apt-get upgrade openssl
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
openssl is already the newest version.
openssl set to manually installed.
The following packages will be upgraded:
  apt apt-utils base-files cpio gcc-4.8-base ifupdown initscripts
  libapt-inst1.5 libapt-pkg4.12 libc-bin libc6 libdrm2 libgcrypt11
  libgnutls-openssl27 libgnutls26 libssl1.0.0 libstdc++6 libudev1 login
  multiarch-support ntpdate passwd perl perl-base perl-modules sysv-rc
  sysvinit-utils udev```

I recreated the container and made an educated guess as to which package was incorrect, ran an apt-get install libssl1.0.0 and sure enough it fixed the problem for curl but not for git, installing libgnutls-openssl27 got git clone working again. The rancher/server container needs to have these packages installed/updated.

Thanks!

**Comment by Seth-Karlo** [comment]
I've continued to debug this since my posting and am no longer certain that this is related to SSLv2/v3 per se, but am still not sure why on certain types of certificates this is failing without the upgraded packages.

**Comment by vincent99** [comment]
The `rancher/server` container is based on Ubuntu 14.04.4 LTS, which packages OpenSSL `1.0.1f` from 2014, plus backported security fixes.  CoreOS has bleeding-edge-ish everything.  One person's old is another's "stable".

It seems somewhat unlikely that the [minor/security changes there have been](http://changelogs.ubuntu.com/changelogs/pool/main/o/openssl/openssl_1.0.1f-1ubuntu2.18/changelog) since 14.04.4 (ubuntu2.16) would fix this.  Is a server with a problem cert available on the internet somewhere for us to hit?  (You can email me privately if you want)

**Comment by Seth-Karlo** [comment]
Unfortunately not, it's our internal private github. I admit my understanding of SSL/TLS and the various versions is somewhat limited, and a colleague of mine has suggested I have it backwards, and that the issue is that the SSL on this will _only_ accept TLS version 1.2, and refuse downgrading to SSL v2/v3. I'm going to do some additional debugging and try to understand in more detail why the package updates fix this. I'll update this thread with any developments.

**Comment by WillPlatnick** [comment]
@Seth-Karlo Starting with the TLS_FALLBACK_SCSV patch in OpenSSL, you can't downgrade SSL protocols.

**Comment by phrag** [comment]
**Problem**
OS is missing "Thawte Premium Server CA"

**Thawte Premium Server CA**
> Country = ZA
Organization = Thawte Consulting cc
Organizational Unit = Class 3 Public Primary Certification Authority - G2
Organizational Unit = Certification Services Division
Common name = Thawte Premium Server CA
Serial Number: 36 12 22 96 c5 e3 38 a5 20 a1 d2 5f 4c d7 09 54
Valid From: Wednesday, July 31, 1996
Valid to: Friday, January 01, 2021
Certificate SHA1 Fingerprint: e0 ab 05 94 20 72 54 93 05 60 62 02 36 70 f7 cd 2e fc 66 66
Key Size: RSA(1024 Bits)
Signature Algorithm: sha1RSA

https://www.thawte.com/roots/thawte_Premium_Server_CA.pem

**Fix**
`wget https://www.thawte.com/roots/thawte_Premium_Server_CA.pem && 
sudo mv thawte_Premium_Server_CA.pem /usr/local/share/ca-certificates/thawte_Premium_Server_CA.crt && 
sudo update-ca-certificates ; `

**Comment by phrag** [comment]
To clarify, TLS1.2 is negotiated, no SSLv2 / SSLv3 ;)

**Comment by vincent99** [comment]
It sounds like you're missing an intermediary cert that the service should be sending; that CA is 1024 bits and not directly trusted by modern clients, but they issue certs with an intermediate signed by it to support older clients.

**Comment by deniseschannon** [comment]
If you are still facing this issue, please re-open it.
